### ISSUE TITLE
SLPVectorizer incorrectly reorders select operands

### ISSUE THREAD

**Issue by shao-hua-li** [llvm/llvm-project#66176]
clang at -O2/3 produced the wrong code.

Bisected to d01aec4c769d50fb92e86decd41d077c94105841, which was committed by @nikic 

Compiler explorer: https://godbolt.org/z/9bhhc5rGM

```console
% cat a.c
int printf(const char *, ...);
int a, c;
long b, g;
static int d[1] = {5};
unsigned h;
long *i() {
  c = 0;
  for (; c <= 2; c++) {
    for (;;) {
      char e[1] = {97};
      char f = *e + c;
      
      if (d[f - 97]) {
        int *j[] = {&d[0], &d[0]};
        h++;
        break;
      }
      return &b;
    }
    for (; h <= 2;)
      return &g;
  }
  for (;;)
    ;
}
int main() {
  i();
  printf("%d\n", a);
}
%
% clang -O0 a.c && ./a.out
0
% clang -O2 a.c
% ./a.out
(Timeout)
%
```

**Comment by nikic** [comment]
The InstCombine patch exposes a poison-propagation issue that has been introduced by some earlier transform. Before the problematic InstCombine run we already have this:
```
  %op.rdx = select i1 poison, i1 true, i1 %cmp6
  br i1 %op.rdx, label %return, label %for.cond12
```

**Comment by nikic** [comment]
Miscompile is in SLPVectorizer (https://llvm.godbolt.org/z/e593Y5r7s):
```llvm
define i1 @src(i32 %inc) {
  %cmp = icmp ult i32 %inc, 3
  %sel1 = select i1 %cmp, i1 true, i1 poison
  %sel2 = select i1 %sel1, i1 true, i1 poison
  %sel3 = select i1 %sel2, i1 true, i1 poison
  ret i1 %sel3
}

define i1 @tgt(i32 %inc) {
  %cmp = icmp ult i32 %inc, 3
  %op.rdx = select i1 poison, i1 true, i1 %cmp
  ret i1 %op.rdx
}
```

The operands of the logical or get reordered incorrectly here.

**Comment by alexey-bataev** [comment]
Fixed in b9ad72ba056126d462d229877cdce1efa0761136

**Comment by nikic** [comment]
Thanks!
